tensorflow/contrib/opt/python/training/model_average_optimizer.py

```diff
@@ -0,0 +1,693 @@
+    self._tokens_per_step = max(total_num_replicas, replicas_to_aggregate)
+    self._global_step = None
+    self._sync_token_queue = None
+
+    # The synchronization op will be executed in a queue runner which should
+    # only be executed by one of the replicas (usually the chief).
+    self._chief_queue_runner = None
+
+    # Remember which accumulator is on which device to set the initial step in
+    # the accumulator to be global step. This list contains list of the
+    # following format: (accumulator, device).
+    self._accumulator_list = []
+    # ModelAverageHook should be called before ReplicasHook.
+    self._ma_run_hook = False
+    # name: string. Name of the global variables and related operation on ps.
+    self._name = 'modelAverage'
+
+    self.generate_local_and_global_variables()
+
+  def generate_local_and_global_variables(self):
+    # Change all variables to local variables.
+    for v in variables.global_variables():
+      if v.op.name.find(self._name) != -1:
+        raise AssertionError('%s: cannot use \'%s\' as a substr of any name for'
+                             ' ops or variables when calling an '
+                             'ModelAverageOptimizer.' % (v.op.name, self._name))
+      ops.add_to_collection(ops.GraphKeys.LOCAL_VARIABLES, v)
+    # Clear global_variables list.
+    ops.get_default_graph().clear_collection(ops.GraphKeys.GLOBAL_VARIABLES)
+    # Generate new global variables dependent on trainable variables.
+    for i, v in enumerate(variables.trainable_variables()):
+      if v.op.name.find(self._name) != -1:
+        raise AssertionError('%s: cannot use \'%s\' as a substr of a name for '
+                             'any ops or variables when calling an '
+                             'ModelAverageOptimizer.' % (v.op.name, self._name))
+      # v_g is the global-variable version of each user-defined trainable
+      # variable.  They are supposed to be placed on PS device. v_g is used to
+      # store averaged model parameters.
+      v_g = variable_scope.get_variable(name='%s_g%d' % (self._name, i),
+                                  initializer=v.initialized_value(),
+                                  trainable=False,
+                                  collections = [ops.GraphKeys.GLOBAL_VARIABLES,
+                                                 'global_model'])
+      v_block_grad = variable_scope.get_variable(
+                  name='%s_block_grad%d' % (self._name, i), shape=v.get_shape(),
+                  initializer=init_ops.constant_initializer(0.0, dtype=v.dtype),
+                  trainable=False,
+                  collections = [ops.GraphKeys.GLOBAL_VARIABLES, 'block_grads'])
+      v_nesterov = variable_scope.get_variable(
+                  name='%s_nesterov%d' % (self._name, i),
+                  initializer=v_g.initialized_value(),
+                  trainable=False,
+                  collections = [ops.GraphKeys.GLOBAL_VARIABLES, 'nesterov'])
```

Comment:
I am still unable to run your code due to an exception that originates here:

> InvalidArgumentError (see above for traceback): Cannot colocate nodes 'modelAverage_nesterov313/cond_627/read/Switch_modelAverage_nesterov313_0' and 'modelAverage_g313: Cannot merge devices with incompatible jobs: '/job:ps/task:3' and '/job:worker'
> [[Node: modelAverage_nesterov313/cond_627/read/Switch_modelAverage_nesterov313_0 = Switch[T=DT_FLOAT, _class=["loc:@modelAverage_g313"], _device="/job:worker"](modelAverage_nesterov313/cond/Merge, cond_627/pred_id)]]

My suspicion is that the device_setter is attempting to place it on the local worker, but the is failing because v_g.initialized_value() is on the ps? I'm not sure.

Any ideas?